### PR TITLE
feat(server): Delay serving global config in managed mode before upstream response

### LINKED ISSUE #2584: Make global config required for processing
Before a global config is loaded / if a global config cannot be loaded during an incident, we currently fall back to using a default global config everywhere.

There are several use cases where if we actually want to rely on dynamic configuration, we have to be able to load it, e.g.

* When we move span metrics extraction config to sentry,
* when we move config for generic inbound filters to sentry.

## Solution

Just like envelopes get buffered until a valid project state is available, we should queue envelopes until a valid global config is available.

### PR DESCRIPTION
**Pull request by TBS1996**
## Problem
currently, before we request global config from upstream, we will simply serve the default global config. This kind of implicit behaviour is not ideal. 

## Solution
This PR will  affect the behaviour in managed mode, where we won't start processing envelopes until we have received a valid global config. To do this we hook into the existing buffer logic in projectcache where we buffer until we receive a projectstate for an envelope, and simply add global config as a second requirement. 

## Implementation

* we move the logic for getting global config from the envelopeprocessor to projectcache, and we include it in the message we send to envelopeprocessor for processing, which guarantees there's a valid global config at time of processing.
* When we receive an envelope, we check if we have a global config `handle_validate_envelope`, if not we buffer the envelope.  
* When we receive projectstates, this used to always trigger a dequeue from the buffer, but now if we don't have global config we stop dequeing and push the projectkey that would have been dequeued into a BTreeSet. 
* When we receive a global config, we request new projectstates for all of the projectkeys in the BTreeSet, which will trigger a new dequeue when their projectstates arrive, this time it should dequeue successfully as we have a global config.

The reason we don't dequeue directly when we receive the global config is because the projectstates might have expired by the time the global config arrives. This might lead to some redundant fetching of projectstates but i figured it would be rather insignificant and it will only happen once, the added complexity of checking was not worth it imo. If you disagree, let me know!

### CODE REVIEW

File: CHANGELOG.md
Comment (olksdr) [inline CHANGELOG.md]:
```suggestion
- Postpone processing till the global config is available. ([#2697](https://github.com/getsentry/relay/pull/2697))
```

File: relay-server/src/actors/global_config.rs
Comment (olksdr) [inline relay-server/src/actors/global_config.rs]: But this is not a message, it's just a wrapper over the global config, isn't it? 

```suggestion
/// Describes the current fetching status of the [`GlobalConfig`] from the upstream.
```

File: relay-server/src/actors/global_config.rs
Comment (olksdr) [inline relay-server/src/actors/global_config.rs]:
```suggestion
    /// Global config ready to be used by other services.
```

File: relay-server/src/actors/global_config.rs
Comment (olksdr) [inline relay-server/src/actors/global_config.rs]:
```suggestion
    /// This variant implies different things in different circumstances. In managed mode, it means
    /// that we have received a config from upstream. In other modes the config is either
    /// from a file or the default global config.
```

File: relay-server/src/actors/global_config.rs
Comment (olksdr) [inline relay-server/src/actors/global_config.rs]:
```suggestion
    /// The global config is requested from the upstream but it has not arrived yet.
```

File: relay-server/src/actors/project_cache.rs
Comment (olksdr) [inline relay-server/src/actors/project_cache.rs]:
```suggestion
    /// directly when we receive [`GlobalConfig`] is that the [`ProjectState`]s might have expired in the meantime,
```

File: relay-server/src/actors/project_cache.rs
Comment (olksdr) [inline relay-server/src/actors/project_cache.rs:472]: Please, add more docs for this enum.

File: relay-server/src/actors/project_cache.rs
Comment (olksdr) [inline relay-server/src/actors/project_cache.rs]:
```suggestion
    /// - [`GlobalConfig`] from the [`GlobalConfigManager`] must be available.
```

File: relay-server/src/actors/project_cache.rs
Comment (olksdr) [inline relay-server/src/actors/project_cache.rs:769]: Why was this changed?

File: relay-server/src/actors/project_cache.rs
Comment (olksdr) [inline relay-server/src/actors/project_cache.rs]:
```suggestion
                                            global_config::Status::Waiting => relay_log::error!("Still waiting for the global config"),
```

File: relay-server/src/endpoints/project_configs.rs
Comment (olksdr) [inline relay-server/src/endpoints/project_configs.rs:138]: So here we always actually return the default if the config currently unavailable?

File: relay-server/src/actors/project_cache.rs
Comment (olksdr) [inline relay-server/src/actors/project_cache.rs]: Why the formatting looks a bit off? 

And also wondering if we still want to be biased for processing the incoming messages first.

File: relay-server/src/actors/project_cache.rs
Comment (olksdr) [inline relay-server/src/actors/project_cache.rs:537]: As I mentioned before, this potentially can be harmful , we can potentially accumulate a lot of data, in case we never can get the proper global config from the upstream.

**Review by olksdr — DISMISSED** [review]
Overall approach looks ok to me so far. And it would be great to add actually the integration test, to check the exact behaviour of the service we expect to see - like we should not process anything till the global config is available and fetched by the Relay.

File: relay-server/src/actors/project_cache.rs
Comment (TBS1996) [reply relay-server/src/actors/project_cache.rs:769]: I think the code is way more clear this way, not related to PR but just following the principle of cleaning up things in the same module I'm working in as i go along.

File: relay-server/src/actors/project_cache.rs
Comment (TBS1996) [reply relay-server/src/actors/project_cache.rs]: i thought my rustformatter was broken so i'd wait a bit later until it does, but maybe it's not supposed to work within a macro ? I'll fix it manually now

File: relay-server/src/actors/project_cache.rs
Comment (TBS1996) [reply relay-server/src/actors/project_cache.rs:537]: I had that in mind when making the recent change as a set of projectkeys should take a lot less space than a vector of messages 

i dont think the memory should be a problem here though, we already have many growing vectors in the projectcachebroker, although maybe we can use the indexed keys to fetch new states when the global config arrives. 

we could sync on this in case we end up going back-and-forth a lot here

File: relay-server/src/endpoints/project_configs.rs
Comment (TBS1996) [reply relay-server/src/endpoints/project_configs.rs:138]: yes, not because i think it's good but because It's what we currently do and I wasn't sure to add another behaviour in the scope of this PR

**Comment by TBS1996** [comment]
>Overall approach looks ok to me so far. And it would be great to add actually the integration test, to check the exact behaviour of the service we expect to see - like we should not process anything till the global config is available and fetched by the Relay.


I definitively agree we need integration test for this, I thought i'd wait until later but I can start right away since the purpose of the integration test is the behaviour not the implementation

File: relay-server/src/actors/global_config.rs
Comment (iker-barriocanal) [inline relay-server/src/actors/global_config.rs:133]: Is an `Option` a good replacement for `Status`? If `None`, no global config is available; if `Some(Arc<GlobalConfig>)`, there's a global config available and ready to be used. I think an `Option` is simpler to understand than introducing a new enum.

File: relay-server/src/actors/project_cache.rs
Comment (iker-barriocanal) [inline relay-server/src/actors/project_cache.rs:999]: nit: could we align these `}` with their opening brackets?

File: relay-server/src/endpoints/project_configs.rs
Comment (iker-barriocanal) [reply relay-server/src/endpoints/project_configs.rs:138]: Does this mean that PoP relays can get a global config and process envelopes when they should not? A processing relay that hasn't fetched a global config from Sentry will send the default config to PoP relays, and PoPs will start processing. This is also the case for the following relays in the chain.

**Review by iker-barriocanal — COMMENTED** [review]
Approach looks good!

File: relay-server/src/actors/global_config.rs
Comment (TBS1996) [reply relay-server/src/actors/global_config.rs:133]: Personally I find custom enum like this more easy to understand, i dont think it's obvious that a None should eventually become a Some. 

But if most of team agree with you then i'll change it to option

File: relay-server/src/endpoints/project_configs.rs
Comment (TBS1996) [reply relay-server/src/endpoints/project_configs.rs:138]: For others: me and iker synced and concluded that it's not a problem in practice because we deploy processing before pops with a significant time delay, and by that time processing should have a global config ready.

please others come with feedback about this

File: relay-server/src/actors/project_cache.rs
Comment (TBS1996) [reply relay-server/src/actors/project_cache.rs:999]: yeah fixed, im not a very good rustformatter X)

File: relay-server/src/actors/global_config.rs
Comment (TBS1996) [inline relay-server/src/actors/global_config.rs:266]: we replace `send` with `send_replace` because in case theres not yet a receiver `send_replace` will populate the receiver when it gets constructed later.

File: relay-server/src/actors/processor.rs
Comment (TBS1996) [inline relay-server/src/actors/processor.rs:323]: moved global config from the envelopeprocessor to the processorstate

File: relay-server/src/actors/processor.rs
Comment (TBS1996) [inline relay-server/src/actors/processor.rs:544]: now that 'inner' is irrelevant, i'll remove it in following pr

File: tests/integration/test_query.py
Comment (TBS1996) [inline tests/integration/test_query.py:145]: I can look deeper into it but I believe it simply needs more retries because after the first 2 retries it have to wait for a bit before it gets the global config and can proceed

File: relay-server/src/actors/global_config.rs
Comment (jjbayer) [reply relay-server/src/actors/global_config.rs:133]: I agree that explicit & documented enums make the code easier to understand.

File: relay-server/src/actors/global_config.rs
Comment (jjbayer) [inline relay-server/src/actors/global_config.rs]: nit: Should we call this `Pending` to align with naming elsewhere in the code?

```suggestion
    Pending,
```

Comment (TBS1996) [reply relay-server/src/actors/global_config.rs]: Agree about the pending name!

File: relay-server/src/actors/global_config.rs
Comment (iker-barriocanal) [inline relay-server/src/actors/global_config.rs:235]: I find this comment useful and I'd suggest restoring it.

File: relay-server/src/actors/project_cache.rs
Comment (iker-barriocanal) [inline relay-server/src/actors/project_cache.rs]:
```suggestion
    fn global_config(&self) -> Option<Arc<GlobalConfig>> {
```
?

File: tests/integration/test_query.py
Comment (iker-barriocanal) [reply tests/integration/test_query.py:145]: Why do project configs need more retries with this PR?

File: tests/integration/test_envelope.py
Comment (iker-barriocanal) [inline tests/integration/test_envelope.py]: Could we move this comment as a docstring of the test?

File: tests/integration/test_envelope.py
Comment (iker-barriocanal) [inline tests/integration/test_envelope.py]: What is the second argument, `None`, for?

File: tests/integration/test_envelope.py
Comment (iker-barriocanal) [inline tests/integration/test_envelope.py:605]: If we set the value to `0`, does it speed the test up? Does this enable reducing the wait time at the bottom of the test?

File: tests/integration/test_envelope.py
Comment (iker-barriocanal) [inline tests/integration/test_envelope.py]: See [`pytest.raises`](https://docs.pytest.org/en/stable/reference/reference.html#pytest.raises).

It seems a bit fishy that we only wait 1s to check if there are events when there aren't supposed to be, and wait 10s when they are supposed to be. Could we be running into false negatives?

File: relay-server/src/actors/global_config.rs
Comment (TBS1996) [reply relay-server/src/actors/global_config.rs:235]: we now use send_replace, which never fails, so the old doc is outdated

File: relay-server/src/actors/project_cache.rs
Comment (TBS1996) [reply relay-server/src/actors/project_cache.rs]: yeahh makes sense,  i went by [this](https://doc.rust-lang.org/std/ops/enum.ControlFlow.html#method.continue_value) precedent but i think it doesnt apply in the same manner here actually.

File: tests/integration/test_envelope.py
Comment (TBS1996) [reply tests/integration/test_envelope.py]: copied from an example, seems i dont need, thx!

File: tests/integration/test_envelope.py
Comment (TBS1996) [reply tests/integration/test_envelope.py:605]: yeah thanks! 

i had forgotten to reduce the 10 sec interval after i reduced the fetch_interval, it would work with 2 sec timeout but now it works with 1 sec after putting the fetch_interval to 0 ^^

File: tests/integration/test_envelope.py
Comment (TBS1996) [reply tests/integration/test_envelope.py:615]: hmm, trying but it doesn't seem to work, using it with assertionerror because thats what gets logged in mini_sentry.test_failures:

>[('/api/666/envelope/', AssertionError('Relay sent us event: global config missing in upstream response')), ('/api/666/envelope/', AssertionError('Relay sent us event: global config missing in upstream response')), ('/api/666/envelope/', AssertionError('Relay sent us event: error fetching project states: upstream request returned error 500 Internal Server Error: no error details')), ('/api/666/envelope/', AssertionError('Relay sent us event: global config missing in upstream response'))]

File: tests/integration/test_query.py
Comment (TBS1996) [reply tests/integration/test_query.py:145]: we trigger another dequeue by asking for fresh projectstates

https://github.com/getsentry/relay/pull/2697/files#diff-ed683d4a7b32e65843545cca6a8d6acd8d9e6779fb4df8cd6153bf9b8b844eb9R476

File: CHANGELOG.md
Comment (olksdr) [inline CHANGELOG.md]: Please, merge the `master` in and move this record to `unreleased` section.

File: relay-server/src/actors/project_cache.rs
Comment (olksdr) [inline relay-server/src/actors/project_cache.rs]: Maybe something like this will be better. 

```suggestion
/// Describes the current status of the [`GlobalConfig`], either it's ready to be used or contains the list of in-flight project keys, to be processed once the config arrives. 
```

File: relay-server/src/endpoints/project_configs.rs
Comment (olksdr) [reply relay-server/src/endpoints/project_configs.rs:138]: The time delay does not matter, if the processing Relays cannot get the global config from upstream. 
When the PoPs are deployed while there is an incident with fetching global config, PoPs will get default one.

File: tests/integration/test_envelope.py
Comment (olksdr) [reply tests/integration/test_envelope.py]: Why this resolved and not addressed?

File: tests/integration/test_envelope.py
Comment (olksdr) [inline tests/integration/test_envelope.py]: Please, send more than 1 envelope, 3 - 5 would do.

File: tests/integration/test_envelope.py
Comment (olksdr) [reply tests/integration/test_envelope.py]: Was this answered or addressed ?

File: tests/integration/test_envelope.py
Comment (olksdr) [inline tests/integration/test_envelope.py]: Please, assert that after the global config is fetched we get all sent envelopes here.

File: relay-server/src/endpoints/project_configs.rs
Comment (TBS1996) [reply relay-server/src/endpoints/project_configs.rs:138]: youre right, its not ideal. I discussed this with @Dav1dde  who also pointed this out. We would have to communicate to the downstream relays that it's a global config they're receiving, or not send back at all. Backwards compatibility is an issue here.

File: relay-server/src/actors/project_cache.rs
Comment (TBS1996) [reply relay-server/src/actors/project_cache.rs]: i agree, thanks!

File: tests/integration/test_envelope.py
Comment (TBS1996) [reply tests/integration/test_envelope.py]: do you mean to do it in a more explicit manner? because afaict, calling get_event does assert that we got the envelope, since it will panic if not present

**Review by olksdr — COMMENTED** [review]
Looks fine to me

You mentioned  you might want to wait on @jan-auer  to give a final review.  Otherwise I would carefully test it in prod.

One important issue here to handle is upstream Relays, to make sure they receive a correct global config. It can be done in the immediate followup PR. But  to be honest, I would make sure it's finished in this PR, so the feature is fully complete.

File: tests/integration/test_envelope.py
Comment (iker-barriocanal) [inline tests/integration/test_envelope.py]: IIUC, this runs without any time delay. I have the same concern as in https://github.com/getsentry/relay/pull/2697#discussion_r1394280536.

**Review by iker-barriocanal — COMMENTED** [review]
Code LGTM, one open question before approving

File: tests/integration/test_envelope.py
Comment (Dav1dde) [reply tests/integration/test_envelope.py]: iirc it has an internal 1 second timeout

File: relay-server/src/endpoints/project_configs.rs
Comment (Dav1dde) [reply relay-server/src/endpoints/project_configs.rs:138]: I think this still mirrors current behaviour, we can tackle this in a separate issue imo. The possible solutions are actually quite tricky and depends on wether we want to change the API or not.

File: tests/integration/test_envelope.py
Comment (Dav1dde) [reply tests/integration/test_envelope.py]: Actually I am wrong, it does **not** have an internal timeout of 1 second. 🚨

Comment (TBS1996) [reply tests/integration/test_envelope.py]: yeah good catch! I added 2 sec sleep now

File: tests/integration/test_envelope.py
Comment (TBS1996) [reply tests/integration/test_envelope.py]: @olksdr yeah we now wait when asserting empty here

**Comment by TBS1996** [comment]
> Looks fine to me
> 
> You mentioned you might want to wait on @jan-auer to give a final review. Otherwise I would carefully test it in prod.
> 
> One important issue here to handle is upstream Relays, to make sure they receive a correct global config. It can be done in the immediate followup PR. But to be honest, I would make sure it's finished in this PR, so the feature is fully complete.

do you mean the issue of them potentially getting a default global config and thinking its from sentry? how to handle that case isnt obvious, so we should probably discuss this together properly before doing anything here

File: tests/integration/test_envelope.py
Comment (TBS1996) [inline tests/integration/test_envelope.py:629]: this one doesn't require sleeping first, since we assert that it's not empty in the previous lines, and we just wanna check that we take out exactly `envelope_qty` envelopes

File: tests/integration/test_envelope.py
Comment (Dav1dde) [reply tests/integration/test_envelope.py]: @TBS1996 `assert_empty` takes a timeout parameter, just use that

Comment (TBS1996) [reply tests/integration/test_envelope.py]: that makes sense, thx! changed it

**Review by olksdr — APPROVED** [review]
Good to test.

The proper propagation of the global config to PoPs can be handled in the followups.

File: tests/integration/test_envelope.py
Comment (iker-barriocanal) [inline tests/integration/test_envelope.py:600]: What's the reason for this change?

This means `global` will be a key with a `None` value. The previous implementation, and Sentry's behavior, is not to include `global`. As a result, this test emulates a different behavior from what Sentry does.

**Pull request merged**
